core/impl/src/main/java/com/blazebit/persistence/impl/keyset/KeysetManager.java

```diff
@@ -57,15 +68,167 @@ public void initialize(List<OrderByExpression> orderByExpressions) {
+
+                renderingBuffer.setLength(0);
+                queryGenerator.generate(orderByExpression.getExpression());
+                String renderedOrderByExpr = renderingBuffer.toString();
+
+                renderingBuffer.setLength(0);
+                applyKeysetParameter(renderingBuffer, i, key[i]);
+                String renderedKeysetParameter = renderingBuffer.toString();
+
+                if (orderByExpression.isDescending() && keysetMode != KeysetMode.PREVIOUS) {
+                    leftHandsideExpressions.add(renderedOrderByExpr);
+                    rightHandsideExpressions.add(renderedKeysetParameter);
+                } else {
+                    leftHandsideExpressions.add(renderedKeysetParameter);
+                    rightHandsideExpressions.add(renderedOrderByExpr);
+                }
+            }
+
+            renderRowValueComparison(sb, keysetMode == KeysetMode.SAME ? "<=" : "<", leftHandsideExpressions, rightHandsideExpressions);
+        }
+    }
+
+    private void renderRowValueComparison(StringBuilder sb, String operator, Collection<String> leftHandsideExpressions, Collection<String> rightHandsideExpressions) {
+        sb.append(jpaProvider.getCustomFunctionInvocation(RowValueConstructorComparisonFunction.FUNCTION_NAME, leftHandsideExpressions.size() + rightHandsideExpressions.size() + 1))
+                .append('\'').append(operator).append('\'');
+        renderArguments(sb, leftHandsideExpressions);
+        renderArguments(sb, rightHandsideExpressions);
+        sb.append(") = true");
+    }
+
+    private void renderArguments(StringBuilder sb, Iterable<String> arguments) {
+        Iterator<String> expressionIterator = arguments.iterator();
+        if (expressionIterator.hasNext()) {
+            sb.append(',');
+            sb.append(expressionIterator.next());
+            while (expressionIterator.hasNext()) {
+                sb.append(',').append(expressionIterator.next());
+            }
+        }
+    }
+
     public void buildKeysetPredicate(StringBuilder sb) {
-        int expressionCount = orderByExpressions.size();
         KeysetLink keysetLink = getKeysetLink();
         KeysetMode keysetMode = keysetLink.getKeysetMode();
         Keyset keyset = keysetLink.getKeyset();
         Serializable[] key;
 
         key = keyset.getTuple();
 
+        buildKeysetPredicate0(keysetMode, key, sb, orderByExpressions);
+    }
+
+    private void buildOptimizedPredicate0(KeysetMode keysetMode, Serializable[] key, StringBuilder sb, List<OrderByExpression> orderByExpressions) {
+        int expressionCount = orderByExpressions.size();
+        int brackets = 1;
+        sb.append('(');
+
+        SimpleQueryGenerator.BooleanLiteralRenderingContext oldBooleanLiteralRenderingContext = queryGenerator.setBooleanLiteralRenderingContext(SimpleQueryGenerator.BooleanLiteralRenderingContext.CASE_WHEN);
+
+        for (int i = 1; i < expressionCount; i++) {
```

Comment:
Would have been nice to have such a comment for the loop starting on line 243 :)